scripts/zones/Aht_Urhgan_Whitegate/npcs/Kaduru-Haiduru.lua

```diff
@@ -0,0 +1,125 @@
+    local shihuDanhuDate = player:getVar("ShihuDanhu_TP_date");
+    local timesUsed = player:getVar("Kaduru_TimesUsed");
+
+    if ((canUse_KaduruHaiduru_Service(player) == 1) and (timesUsed == 3)) then
+        if (trade:getItemCount() == 1) then
+            if (trade:hasItemQty(2185, 1)) then
+                player:startEvent(0x009A, 0, player:getNation(), 0, 0, 0, 0, 0, 0, 0);
+                player:setVar("ShihuDanhu_TP_date", 0);
+                player:setVar("Kaduru_ShihuDanhu_date", 0);
+            end
+        end
+    elseif ((canUse_KaduruHaiduru_Service(player) == 1) and (timesUsed < 3)) then
+        if (trade:getItemCount() == 1) then
+            if (trade:hasItemQty(2185, 1)) then
+                player:startEvent(0x0098, 0, 0, 0, 0, 0, 0, 0, 0, 0);
+                player:setVar("ShihuDanhu_TP_date", 0);
+                player:setVar("Kaduru_ShihuDanhu_date", 0);
+            end
+        end
+    else
+        if (caughtUsingShihuDanhuDate == 0) then
+            player:setVar("Kaduru_ShihuDanhu_date", os.date("%j"));
+            player:setVar("Kaduru_TimesUsed", 0);
+        end
+        player:startEvent(0x009B, 0, 0, 0, 0, 0, 0, 0, 0, 0);
+    end
+end
+
+-----------------------------------
+-- onEventUpdate
+-----------------------------------
+function onEventUpdate(player, csid, option)
+    --printf("CSID: %u",csid);
+    --printf("RESULT: %u",option);
+end;
+
+-----------------------------------
+-- onEventFinish
+-----------------------------------
+function onEventFinish(player, csid, option)
+    --printf("CSID: %u",csid);
+    --printf("RESULT: %u",option);
+    local timesUsed = player:getVar("Kaduru_TimesUsed");
+    if (csid == 0x009A) then
+        if (option == 1) then       -- Duchy of Jeuno
+            player:setPos(0, 3, -6, 190, 243);
+        elseif (option == 2) then   -- Nation of Allegiance
+            local nation = player:getNation();
+            switch (nation): caseof
+            {
+                [0] = function(x)  -- San d'Oria
+                    player:setPos(110, 0, -7, 175, 231);
+                end,
+                [1] = function(x)  -- Bastok
+                    player:setPos(90, 0, -67, 248, 234);
+                end,
+                [2] = function(x)  -- Windurst
+                    player:setPos(192, -12, 218, 60, 240);
+                end,
+            }
+        end
+        if (timesUsed < 3) then
+            player:setVar("Kaduru_TimesUsed", (timesUsed + 1));
+        end
+        player:tradeComplete();
+    elseif (csid == 0x0098) then    -- Duchy of Jeuno only
+        -- We have either not used Kaduru-Haiduru enough, or we were reset after using Shihu-Danhu.
+        player:setPos(0, 3, -6, 190, 243);
+        if (timesUsed < 3) then
+            player:setVar("Kaduru_TimesUsed", (timesUsed + 1));
+        end
+        player:tradeComplete();
+    end
+end;
+
+function canUse_KaduruHaiduru_Service(player)
```

Comment:
I think it was wren that showed me we can scope these local if we place them above instead of below.
`local function blahblah()`